public/app/core/utils/explore.ts

```diff
@@ -474,23 +474,23 @@ export const getRefIds = (value: any): string[] => {
 };
 
 export const sortInAscendingOrder = (a: LogRowModel, b: LogRowModel) => {
-  if (a.timestamp < b.timestamp) {
+  if (a.timeEpochMs < b.timeEpochMs) {
```

Comment:
This was comparing strings before, which works but is kinda weird. One thing though is that this is ms precision right now (was also before) so not sure if it can mess up some ordering for loki where ns precision is default. It is an issue in case of merging multiple streams so there super exact ordering probably isn't important.

Reply:
> One thing though is that this is ms precision right now (was also before)

Was it really? IIRC the string comparison here was done because it was in ns.

Reply:
At some point this was changed https://github.com/grafana/grafana/blob/9ebd99d595bb35cadad5c545878d1165d5cbd712/public/app/plugins/datasource/loki/result_transformer.ts#L89. I assume because of new loki version sends the tiemstamp as string epoch ns which isn't parseable by dateTime (while ns precision ISO string is just throws away the precision).

Reply:
So right now this is kind of least common denominator approach. This way we do not have to assume the format and parse-ability in explore code. At the same time we still pass on the full dataframe so datasources can keep the raw timestamp and use it later if they get passed a row (like in the case of the context)

Reply:
> This was comparing strings before, which works but is kinda weird

this was mainly comment for recent code where you could get ISO string with either ns or ms precision (depending on datasource) which again works but it wasn't that clear imho what is going.